source/common/grpc/async_client_impl.h

```diff
@@ -0,0 +1,215 @@
+      return nullptr;
+    }
+
+    grpc_stream->set_stream(http_stream);
+
+    Http::MessagePtr message = Common::prepareHeaders(
+        remote_cluster_name_, service_method.service()->full_name(), service_method.name());
+    callbacks.onCreateInitialMetadata(message->headers());
+
+    http_stream->sendHeaders(message->headers(), false);
+    grpc_stream->moveIntoList(std::move(grpc_stream), active_streams_);
+
+    return active_streams_.front().get();
+  }
+
+private:
+  Upstream::ClusterManager& cm_;
+  const std::string remote_cluster_name_;
+  std::list<std::unique_ptr<AsyncClientStreamImpl<RequestType, ResponseType>>> active_streams_;
+
+  friend class AsyncClientStreamImpl<RequestType, ResponseType>;
+};
+
+template <class RequestType, class ResponseType>
+class AsyncClientStreamImpl : public AsyncClientStream<RequestType>,
+                              Http::AsyncClient::StreamCallbacks,
+                              LinkedObject<AsyncClientStreamImpl<RequestType, ResponseType>> {
+public:
+  AsyncClientStreamImpl(AsyncClientImpl<RequestType, ResponseType>& parent,
+                        AsyncClientCallbacks<ResponseType>& callbacks)
+      : parent_(parent), callbacks_(callbacks) {}
+
+  // Http::AsyncClient::StreamCallbacks
+  void onHeaders(Http::HeaderMapPtr&& headers, bool end_stream) override {
+    if (remote_closed_) {
+      return;
+    }
+    if (Http::Utility::getResponseStatus(*headers) != enumToInt(Http::Code::OK)) {
+      reset();
+      callbacks_.onRemoteClose(GrpcStatus::Internal);
+      return;
+    }
+    if (end_stream) {
+      onTrailers(std::move(headers));
+      return;
+    }
+    callbacks_.onReceiveInitialMetadata(std::move(headers));
+  }
+
+  void onData(Buffer::Instance& data, bool end_stream) override {
+    if (remote_closed_) {
+      return;
+    }
+
+    std::vector<Frame> frames;
+    if (!decoder_.decode(data, frames)) {
+      reset();
+      callbacks_.onRemoteClose(GrpcStatus::Internal);
+      return;
+    }
+
+    for (const auto& frame : frames) {
+      std::unique_ptr<ResponseType> response(new ResponseType());
+      // TODO(htuch): We can avoid linearizing the buffer here when Buffer::Instance implements
+      // protobuf ZeroCopyInputStream.
+      if (frame.flags_ != GRPC_FH_DEFAULT ||
+          !response->ParseFromArray(frame.data_->linearize(frame.data_->length()),
+                                    frame.data_->length())) {
+        reset();
+        callbacks_.onRemoteClose(GrpcStatus::Internal);
+        return;
+      }
+      callbacks_.onReceiveMessage(std::move(response));
+    }
+    if (end_stream) {
+      reset();
+      callbacks_.onRemoteClose(GrpcStatus::Internal);
+      return;
+    }
+  }
+
+  void onTrailers(Http::HeaderMapPtr&& trailers) override {
+    if (remote_closed_) {
+      return;
+    }
+
+    const GrpcStatus grpc_status = Common::getGrpcStatus(*trailers);
+    if (grpc_status == GrpcStatus::InvalidCode) {
```

Comment:
I'm not sure what the ask is here, can you elaborate some more? The status codes are not part of this PR, I temporarily patched in an earlier revision of #1067, which I plan to merge back in here after it hits master. In the most recent #1067, if `grpc_status < 0` in this scenario, it indicates a missing/invalid code, so I will update this logic to reflect that.